In [0]:
import pandas as pd
import numpy as np

# 1. Getting the Data

In [0]:
%sh wget https://www.dropbox.com/s/l6tmmgkdy72j1kp/Fire_Department_Calls_for_Service_sampled_1_percent.csv

--2022-04-02 00:31:12-- https://www.dropbox.com/s/l6tmmgkdy72j1kp/Fire_Department_Calls_for_Service_sampled_1_percent.csv
Resolving www.dropbox.com (www.dropbox.com)... 162.125.5.18, 2620:100:601d:18::a27d:512
Connecting to www.dropbox.com (www.dropbox.com)|162.125.5.18|:443... connected.
HTTP request sent, awaiting response... 301 Moved Permanently
Location: /s/raw/l6tmmgkdy72j1kp/Fire_Department_Calls_for_Service_sampled_1_percent.csv [following]
--2022-04-02 00:31:12-- https://www.dropbox.com/s/raw/l6tmmgkdy72j1kp/Fire_Department_Calls_for_Service_sampled_1_percent.csv
Reusing existing connection to www.dropbox.com:443.
HTTP request sent, awaiting response... 302 Found
Location: https://ucfd46f36ff6a52ed828912a58e3.dl.dropboxusercontent.com/cd/0/inline/BimXhOdvuG32IZDQtKRftJgB9M1RS0bZQzzHYUsXhsojTgy31Sow90PuisOEUxMnSbcGbB8d6GOKmNoXUCs-ZTlh4IF_ZUUQI6oASX4drirneALtuPjwdJnq-D6gVPEO06OdopAQkiAkeXup5BNauSf3odI19ED5haHC0WTLlaE_3w/file# [following]
--2022-04-02 00:31:13-- https://ucfd46f36ff6a52ed828912a58e3.dl.dropboxusercontent.com/cd/0/inline/BimXhOdvuG32IZDQtKRftJgB9M1RS0bZQzzHYUsXhsojTgy31Sow90PuisOEUxMnSbcGbB8d6GOKmNoXUCs-ZTlh4IF_ZUUQI6oASX4drirneALtuPjwdJnq-D6gVPEO06OdopAQkiAkeXup5BNauSf3odI19ED5haHC0WTLlaE_3w/file
Resolving ucfd46f36ff6a52ed828912a58e3.dl.dropboxusercontent.com (ucfd46f36ff6a52ed828912a58e3.dl.dropboxusercontent.com)... 162.125.5.15, 2620:100:601d:15::a27d:50f
Connecting to ucfd46f36ff6a52ed828912a58e3.dl.dropboxusercontent.com (ucfd46f36ff6a52ed828912a58e3.dl.dropboxusercontent.com)|162.125.5.15|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 20895637 (20M) [text/plain]
Saving to: ‘Fire_Department_Calls_for_Service_sampled_1_percent.csv’

 0K .......... .......... .......... .......... .......... 0% 1.50M 13s
 50K .......... .......... .......... .......... .......... 0% 3.56M 9s
 100K .......... .......... .......... .......... .......... 0% 5.40M 7s
 150K .......... .......... .......... .......... .......... 0% 6.79M 6s
 200K .......... .......... .......... .......... .......... 1% 10.1M 5s
 250K .......... .......... .......... .......... .......... 1% 5.63M 5s
 300K .......... .......... .......... .......... .......... 1% 111M 4s
 350K .......... .......... .......... .......... .......... 1% 6.95M 4s
 400K .......... .......... .......... .......... .......... 2% 163M 4s
 450K .......... .......... .......... .......... .......... 2% 18.9M 3s
 500K .......... .......... .......... .......... .......... 2% 25.9M 3s
 550K .......... .......... .......... .......... .......... 2% 23.7M 3s
 600K .......... .......... .......... .......... .......... 3% 26.1M 3s
 650K .......... .......... .......... .......... .......... 3% 21.1M 3s
 700K .......... .......... .......... .......... .......... 3% 39.4M 3s
 750K .......... .......... .......... .......... .......... 3% 24.4M 2s
 800K .......... .......... .......... .......... .......... 4% 21.6M 2s
 850K .......... .......... .......... .......... .......... 4% 38.5M 2s
 900K .......... .......... .......... .......... .......... 4% 28.7M 2s
 950K .......... .......... .......... .......... .......... 4% 32.5M 2s
 1000K .......... .......... .......... .......... .......... 5% 41.9M 2s
 1050K .......... .......... .......... .......... .......... 5% 40.9M 2s
 1100K .......... .......... .......... .......... .......... 5% 42.5M 2s
 1150K .......... .......... .......... .......... .......... 5% 34.9M 2s
 1200K .......... .......... .......... .......... .......... 6% 39.7M 2s
 1250K .......... .......... .......... .......... .......... 6% 39.6M 2s
 1300K .......... .......... .......... .......... .......... 6% 34.3M 2s
 1350K .......... .......... .......... .......... .......... 6% 35.2M 2s
 1400K .......... .......... .......... .......... .......... 7% 34.4M 2s
 1450K .......... .......... .......... .......... .......... 7% 40.7M 1s
 1500K .......... .......... .......... .......... .......... 7% 37.5M 1s
 1550K ........

<b> Specify  the list of column names and the list of datetime columns </b>

In [0]:
col_names = [
    "col1",
    "Call Number",
    "Unit ID",
    "Incident Number",
    "Call Type",
    "Call Date",
    "Watch Date",
    "Received DtTm",
    "Entry DtTm",
    "Dispatch DtTm",
    "Response DtTm",
    "On Scene DtTm",
    "Transport DtTm",
    "Hospital DtTm",
    "Call Final Disposition",
    "Available DtTm",
    "Address",
    "City",
    "Zipcode of Incident",
    "Battalion",
    "Station Area",
    "Box",
    "Original Priority",
    "Priority",
    "Final Priority",
    "ALS Unit",
    "Call Type Group",
    "Number of Alarms",
    "Unit Type",
    "Unit sequence in call dispatch",
    "Fire Prevention District",
    "Supervisor District",
    "Neighborhooods - Analysis Boundaries",
    "Location",
    "RowID",
    "Supervisor Districts",
    "Fire Prevention Districts",
    "Current Police Districts",
    "Neighborhoods - Analysis Boundaries",
    "Zip Codes",
    "Neighborhoods",
    "Police Districts",
    "Civic Center Harm Reduction Project Boundary",
    "HSOC Zones"    
];

dt_col_names = [
    "Received DtTm",
    "Entry DtTm",
    "Dispatch DtTm",
    "Response DtTm",
    "On Scene DtTm",
    "Transport DtTm",
    "Hospital DtTm",
    "Available DtTm",
];

<b> Load the CSV file with datatime parser </b>

In [0]:
dt_parser = lambda x: pd.to_datetime(x, format='%m/%d/%Y %I:%M:%S %p', errors='coerce')

df = pd.read_csv('Fire_Department_Calls_for_Service_sampled_1_percent.csv',
                 sep         = ",", 
                 skiprows    = 1, 
                 names       = col_names, 
                 parse_dates = dt_col_names, 
                 date_parser = dt_parser)

# parse the 'Call Date' column as well
df['Call_Date_Dt'] = pd.to_datetime(df['Call Date'], format='%m/%d/%Y')

# deploy in seconds
df['Delay'] = (df['On Scene DtTm']-df['Received DtTm'])/ np.timedelta64(1, 's')

df.head()

Out[4]:

col1 
 Call Number 
 Unit ID 
 Incident Number 
 Call Type 
 Call Date 
 Watch Date 
 Received DtTm 
 Entry DtTm 
 Dispatch DtTm 
 Response DtTm 
 On Scene DtTm 
 Transport DtTm 
 Hospital DtTm 
 Call Final Disposition 
 Available DtTm 
 Address 
 City 
 Zipcode of Incident 
 Battalion 
 Station Area 
 Box 
 Original Priority 
 Priority 
 Final Priority 
 ALS Unit 
 Call Type Group 
 Number of Alarms 
 Unit Type 
 Unit sequence in call dispatch 
 Fire Prevention District 
 Supervisor District 
 Neighborhooods - Analysis Boundaries 
 Location 
 RowID 
 Supervisor Districts 
 Fire Prevention Districts 
 Current Police Districts 
 Neighborhoods - Analysis Boundaries 
 Zip Codes 
 Neighborhoods 
 Police Districts 
 Civic Center Harm Reduction Project Boundary 
 HSOC Zones 
 Call_Date_Dt 
 Delay 
 
 
 
 
 0 
 2014158 
 123270200 
 T15 
 12108775 
 Other 
 11/22/2012 
 11/22/2012 
 2012-11-22 15:00:45 
 2012-11-22 15:01:07 
 2012-11-22 15:01:14 
 2012-11-22 15:02:36 
 2012-11-22 15:08:10 
 NaT 
 NaT 
 Fire 
 2012-11-22 15:34:51 
 0 Block of BLYTHDALE AVE 
 SF 
 94134.0 
 B09 
 43 
 6244.0 
 3 
 3 
 3 
 False 
 Alarm 
 1 
 TRUCK 
 1.0 
 9 
 10 
 Visitacion Valley 
 (37.710658718384, -122.417699843965) 
 123270200-T15 
 8.0 
 9.0 
 4.0 
 40.0 
 309.0 
 39.0 
 7.0 
 NaN 
 NaN 
 2012-11-22 
 445.0 
 
 
 1 
 3069882 
 82920313 
 93 
 8089738 
 Medical Incident 
 10/18/2008 
 10/18/2008 
 2008-10-18 19:37:11 
 2008-10-18 19:38:19 
 2008-10-18 19:38:30 
 2008-10-18 19:39:34 
 2008-10-18 19:46:03 
 2008-10-18 20:05:33 
 2008-10-18 20:27:14 
 Code 2 Transport 
 2008-10-18 20:55:42 
 2100 Block of JENNINGS ST 
 SF 
 94124.0 
 B10 
 17 
 6651.0 
 3 
 3 
 3 
 True 
 NaN 
 1 
 MEDIC 
 3.0 
 10 
 10 
 Bayview Hunters Point 
 (37.726697750164, -122.38889898772) 
 082920313-93 
 8.0 
 10.0 
 3.0 
 1.0 
 58.0 
 1.0 
 3.0 
 NaN 
 NaN 
 2008-10-18 
 532.0 
 
 
 2 
 4501324 
 22840051 
 E35 
 2084066 
 Alarms 
 10/11/2002 
 10/10/2002 
 2002-10-11 06:31:30 
 2002-10-11 06:31:58 
 2002-10-11 06:32:03 
 NaT 
 2002-10-11 06:33:57 
 NaT 
 NaT 
 Other 
 2002-10-11 06:47:31 
 800 Block of THE EMBARCADERO 
 SF 
 94111.0 
 B03 
 35 
 940.0 
 3 
 3 
 3 
 False 
 NaN 
 1 
 ENGINE 
 1.0 
 1 
 3 
 Financial District/South Beach 
 (37.802567514252, -122.400381980995) 
 022840051-E35 
 10.0 
 3.0 
 1.0 
 8.0 
 28860.0 
 6.0 
 1.0 
 NaN 
 NaN 
 2002-10-11 
 147.0 
 
 
 3 
 4581480 
 21500234 
 E03 
 2044484 
 Structure Fire 
 05/30/2002 
 05/30/2002 
 2002-05-30 13:02:28 
 2002-05-30 13:02:28 
 2002-05-30 13:02:35 
 2002-05-30 13:03:46 
 2002-05-30 13:05:01 
 NaT 
 NaT 
 Other 
 2002-05-30 13:05:22 
 600 Block of JONES ST 
 SF 
 94102.0 
 B01 
 03 
 1462.0 
 3 
 3 
 3 
 False 
 NaN 
 1 
 ENGINE 
 1.0 
 1 
 3 
 Tenderloin 
 (37.787065466391, -122.413159816666) 
 021500234-E03 
 10.0 
 5.0 
 1.0 
 36.0 
 28852.0 
 36.0 
 1.0 
 NaN 
 NaN 
 2002-05-30 
 153.0 
 
 
 4 
 2582789 
 102780322 
 E41 
 10088489 
 Alarms 
 10/05/2010 
 10/05/2010 
 2010-10-05 18:49:42 
 2010-10-05 18:50:19 
 2010-10-05 18:50:40 
 2010-10-05 18:51:44 
 2010-10-05 18:54:25 
 NaT 
 NaT 
 Other 
 2010-10-05 19:00:30 
 500 Block of JONES ST 
 SF 
 94102.0 
 B01 
 03 
 1462.0 
 3 
 3 
 3 
 False 
 NaN 
 1 
 ENGINE 
 2.0 
 1 
 6 
 Tenderloin 
 (37.786268441716, -122.412997235896) 
 102780322-E41 
 9.0 
 5.0 
 10.0 
 36.0 
 28852.0 
 36.0 
 10.0 
 NaN 
 NaN 
 2010-10-05 
 283.0

In [0]:
df.dtypes

Out[5]: col1 int64
Call Number int64
Unit ID object
Incident Number int64
Call Type object
Call Date object
Watch Date object
Received DtTm datetime64[ns]
Entry DtTm datetime64[ns]
Dispatch DtTm datetime64[ns]
Response DtTm datetime64[ns]
On Scene DtTm datetime64[ns]
Transport DtTm datetime64[ns]
Hospital DtTm datetime64[ns]
Call Final Disposition object
Available DtTm datetime64[ns]
Address object
City object
Zipcode of Incident float64
Battalion object
Station Area object
Box float64
Original Priority object
Priority object
Final Priority int64
ALS Unit bool
Call Type Group object
Number of Alarms int64
Unit Type object
Unit sequence in call dispatch float64
Fire Prevention District object
Supervisor District object
Neighborhooods - Analysis Boundaries object
Location object
RowID object
Supervisor Districts float64
Fire Prevention Districts float64
Current Police Districts float64
Neighborhoods - Analysis Boundaries float64
Zip Codes float64
Neighborhoods float64
Police Districts float64
Civic Center Harm Reduction Project Boundary float64
HSOC Zones float64
Call_Date_Dt datetime64[ns]
Delay float64
dtype: object

<b> Create spark dataframe and a view as well </b>

In [0]:
sparkDF=spark.createDataFrame(df) 
sparkDF.printSchema()

root
-- col1: long (nullable = true)
-- Call Number: long (nullable = true)
-- Unit ID: string (nullable = true)
-- Incident Number: long (nullable = true)
-- Call Type: string (nullable = true)
-- Call Date: string (nullable = true)
-- Watch Date: string (nullable = true)
-- Received DtTm: timestamp (nullable = true)
-- Entry DtTm: timestamp (nullable = true)
-- Dispatch DtTm: timestamp (nullable = true)
-- Response DtTm: timestamp (nullable = true)
-- On Scene DtTm: timestamp (nullable = true)
-- Transport DtTm: timestamp (nullable = true)
-- Hospital DtTm: timestamp (nullable = true)
-- Call Final Disposition: string (nullable = true)
-- Available DtTm: timestamp (nullable = true)
-- Address: string (nullable = true)
-- City: string (nullable = true)
-- Zipcode of Incident: double (nullable = true)
-- Battalion: string (nullable = true)
-- Station Area: string (nullable = true)
-- Box: double (nullable = true)
-- Original Priority: string (nullable = true)
-- Priority: string (nullable = true)
-- Final Priority: long (nullable = true)
-- ALS Unit: boolean (nullable = true)
-- Call Type Group: string (nullable = true)
-- Number of Alarms: long (nullable = true)
-- Unit Type: string (nullable = true)
-- Unit sequence in call dispatch: double (nullable = true)
-- Fire Prevention District: string (nullable = true)
-- Supervisor District: string (nullable = true)
-- Neighborhooods - Analysis Boundaries: string (nullable = true)
-- Location: string (nullable = true)
-- RowID: string (nullable = true)
-- Supervisor Districts: double (nullable = true)
-- Fire Prevention Districts: double (nullable = true)
-- Current Police Districts: double (nullable = true)
-- Neighborhoods - Analysis Boundaries: double (nullable = true)
-- Zip Codes: double (nullable = true)
-- Neighborhoods: double (nullable = true)
-- Police Districts: double (nullable = true)
-- Civic Center Harm Reduction Project Boundary: double (nullable = true)
-- HSOC Zones: double (nullable = true)
-- Call_Date_Dt: timestamp (nullable = true)
-- Delay: double (nullable = true)

In [0]:
sparkDF.cache();

In [0]:
sparkDF.createOrReplaceTempView("sparkView")

# 2. Quick check on the data

<b> 2.1 Get some distinct counts </b>

In [0]:
print("Number of Rows:",                            sparkDF.count())
print("Number of distinct COL1 column:",            sparkDF.select('col1').distinct().count())
print("Number of distinct ROWID column:",           sparkDF.select('RowID').distinct().count())
print("Number of distinct CALL NUMBER column:",     sparkDF.select('Call Number').distinct().count())
print("Number of distinct INCIDENT NUMBER column:", sparkDF.select('Incident Number').distinct().count())

Number of Rows: 50536
Number of distinct COL1 column: 50536
Number of distinct ROWID column: 50536
Number of distinct CALL NUMBER column: 49948
Number of distinct INCIDENT NUMBER column: 49948

<b> Finding: </b> 
1. <b> RowID </b> is a valid row identifier which has the same distinct count as number of Rows. No duplicated rows.
2. <b> In the original CSV file, the first column has no column name </b>, and this column is not found in data description on the data.sfgov.org. HOWEVER, it looks like this is another ID column. no further investigation is needed.
3. <b> Number of distinct calls (49948) is different than number of distinct records (50536) </b>. Further investigation is needed to make sure this is 1) a data quality issue, OR 2) the same call might have multiple records in the data.

<b> 2.2 Check why Number of distinct calls (49948) is different than number of distinct records (50536) </b>

In [0]:
%sql
select `Call Number`, count(*) as numRowPerCallNumber 
from sparkView 
group by `Call Number`
   having numRowPerCallNumber > 1

Call Number numRowPerCallNumber 101320313 2 141691481 2 83120182 2 120690347 2 70430181 2 162123048 2 130450226 2 180631724 2 13000088 2 20190274 2 71380021 2 80270074 2 162102629 2 90210271 2 12560257 2 161342024 3 171151018 2 181913144 3 62980157 2 122060371 2 71650148 2 11270024 2 131470184 2 43420026 2 150241939 2 63460060 2 122310143 2 2450128 2 90540140 2 153221631 2 163101781 2 83490014 2 140200271 2 130480086 2 80150398 2 152353252 2 130650176 2 52230010 2 31050034 2 61180189 2 31910132 2 152393807 2 62530084 2 112120050 2 100080337 2 62410193 2 33290146 2 91770340 2 51710270 2 183482988 2 71940391 3 140470013 2 93230293 2 133170021 2 172473837 2 172012966 2 163223985 2 31500034 2 172194206 2 2260207 2 13060148 2 42100259 2 130800225 2 12810100 2 31310056 2 190411112 2 101630289 2 71160236 2 73580007 2 10770398 2 22100361 2 92520288 2 101830123 2 51660020 2 112660135 2 30920115 2 70970063 2 133160060 2 51980249 2 181123160 2 91020118 2 172320376 2 182643243 2 20440033 2 122430153 2 61000012 2 51650062 2 110290104 2 32980373 2 112600263 2 182052069 2 131250050 2 91000322 2 101770065 2 11800141 2 102950360 2 61590455 2 180541854 2 141471211 2 120850157 2 190401763 2 73200130 2 152301118 2 121860349 2 83510315 2 163472786 2 91790222 2 92680112 2 112190312 2 81180348 2 143300316 2 181922444 2 42880299 2 42640185 2 182602451 2 63080169 2 162122149 2 42840139 2 130720063 2 191663755 2 91100202 2 82730170 2 152830714 2 103480045 2 61770131 2 150192925 2 90730080 2 12870396 2 181123446 2 180700587 2 142832331 2 83220061 2 153420539 2 31450337 2 183572665 2 21080411 2 152251926 2 170733727 2 153591126 2 92590101 2 181250218 2 52290013 2 12380091 2 113500147 2 71350009 2 12870030 2 173482910 2 152080565 2 110790070 2 190403760 2 40580367 2 151663143 2 191813694 2 83600254 2 191563748 2 111730342 2 12280005 2 83310152 2 70860004 2 62360168 2 23040285 2 91830227 2 41780235 2 151283526 2 72680032 2 80840080 2 61930040 2 101890232 2 160153809 2 32940094 2 92360005 2 160651098 2 183382490 2 60550394 3 122040008 2 31300239 2 3070014 2 30650436 2 123660013 2 33270089 2 163262943 2 33570017 2 151600555 2 130570318 2 81620161 2 40650323 2 131010278 2 82890360 2 163234046 2 152720731 2 42370208 2 12190023 2 80820063 2 42400337 2 42950096 2 130950313 2 30830016 2 152772136 2 61390347 2 100110111 2 132070303 2 62310059 2 92600419 2 80950369 2 13480234 2 182470937 2 110230007 2 90120330 2 130860008 2 20830211 2 111430149 2 32630010 2 53410188 2 191150057 2 131870122 2 70080223 2 83610265 2 103100023 2 120710069 2 22610244 2 153343915 2 60370351 2 112060237 2 180502259 2 122230380 2 92880397 2 131730123 2 132210115 2 80960342 2 182240300 2 140740082 2 31020407 2 111740305 2 43570003 2 112420095 2 21630250 2 151421456 2 172030878 2 102980350 2 120990339 2 30670301 2 113430320 2 153540277 2 60940121 2 62590394 2 81690241 2 42560274 2 70150183 2 60280060 2 100880272 2 132970314 2 190780138 2 170542810 2 93330166 2 40420428 2 20560299 2 160853114 3 71120227 2 121680063 2 141860027 2 3420433 2 153060468 2 190190208 2 63280093 2 151943220 2 150330822 2 190371957 2 132100045 2 92110232 2 172403028 2 151443512 2 93410326 3 80330278 2 171990066 2 93360327 2 51300288 2 163112790 2 12300039 2 181940319 2 153631050 2 153243997 2 123640109 2 130070132 2 173590098 2 82820249 2 103130202 2 10430393 2 60530146 2 73400295 2 142063347 2 101400226 2 42200363 2 151372236 2 62240189 2 100330047 2 52100337 2 113530032 2 40660226 2 80960079 2 183273158 2 40200162 2 31520397 2 112710034 2 121620196 2 1960303 2 151483553 2 152364197 2 10260389 2 42390310 3 100860267 3 161763087 2 62120030 2 163512421 2 70080299 2 22720389 2 161971107 2 161731987 2 60540199 2 140750253 2 10310167 2 52050326 2 93370073 2 171250264 2 60600031 2 192232390 2 103270033 2 130290352 2 122970323 2 103570382 2 140440266 2 1790053 2 103210078 2 23120069 2 173251377 2 22920399 2 22020017 2 50280261 2 160600217 2 72980299 2 42500177 2 50290240 3 30270343 2 161051622 2 60970282 2 172454041 2 12313

In [0]:
%sql
select * from sparkView
where `Call Number` in ("141691481", "140570353", "83120182","120690347")
order by `Call Number`

col1 Call Number Unit ID Incident Number Call Type Call Date Watch Date Received DtTm Entry DtTm Dispatch DtTm Response DtTm On Scene DtTm Transport DtTm Hospital DtTm Call Final Disposition Available DtTm Address City Zipcode of Incident Battalion Station Area Box Original Priority Priority Final Priority ALS Unit Call Type Group Number of Alarms Unit Type Unit sequence in call dispatch Fire Prevention District Supervisor District Neighborhooods - Analysis Boundaries Location RowID Supervisor Districts Fire Prevention Districts Current Police Districts Neighborhoods - Analysis Boundaries Zip Codes Neighborhoods Police Districts Civic Center Harm Reduction Project Boundary HSOC Zones Call_Date_Dt Delay 3056278 83120182 RC2 8095730 Medical Incident 11/07/2008 11/07/2008 2008-11-07T11:43:20.000+0000 2008-11-07T11:44:20.000+0000 2008-11-07T11:44:39.000+0000 null 2008-11-07T11:49:40.000+0000 null null Other 2008-11-07T12:12:19.000+0000 2200 Block of POST ST SF 94115.0 B04 10 4131.0 3 3 3 true null 1 RESCUE CAPTAIN 2.0 5 5 Japantown (37.784370567305, -122.438841200784) 083120182-RC2 11.0 15.0 9.0 15.0 29490.0 13.0 6.0 null null 2008-11-07T00:00:00.000+0000 380.0 3056279 83120182 82 8095730 Medical Incident 11/07/2008 11/07/2008 2008-11-07T11:43:20.000+0000 2008-11-07T11:44:20.000+0000 2008-11-07T11:44:39.000+0000 2008-11-07T11:45:33.000+0000 2008-11-07T11:49:45.000+0000 null null Patient Declined Transport 2008-11-07T12:12:39.000+0000 2200 Block of POST ST SF 94115.0 B04 10 4131.0 3 3 3 true null 1 MEDIC 3.0 5 5 Japantown (37.784370567305, -122.438841200784) 083120182-82 11.0 15.0 9.0 15.0 29490.0 13.0 6.0 null null 2008-11-07T00:00:00.000+0000 385.0 2199947 120690347 T07 12023036 Structure Fire 03/09/2012 03/09/2012 2012-03-09T20:16:13.000+0000 2012-03-09T20:18:02.000+0000 2012-03-09T20:18:52.000+0000 2012-03-09T20:20:02.000+0000 2012-03-09T20:22:07.000+0000 null null Fire 2012-03-09T23:18:13.000+0000 1200 Block of SOUTH VAN NESS AVE SF 94110.0 B06 07 5511.0 3 3 3 false Fire 1 TRUCK 2.0 6 9 Mission (37.753166244582, -122.416408491437) 120690347-T07 7.0 2.0 7.0 20.0 28859.0 19.0 4.0 null 3.0 2012-03-09T00:00:00.000+0000 354.0 2199946 120690347 E11 12023036 Structure Fire 03/09/2012 03/09/2012 2012-03-09T20:16:13.000+0000 2012-03-09T20:18:02.000+0000 2012-03-09T20:18:52.000+0000 2012-03-09T20:20:15.000+0000 2012-03-09T20:22:24.000+0000 null null Other 2012-03-09T22:14:07.000+0000 1200 Block of SOUTH VAN NESS AVE SF 94110.0 B06 07 5511.0 3 3 3 true Fire 1 ENGINE 4.0 6 9 Mission (37.753166244582, -122.416408491437) 120690347-E11 7.0 2.0 7.0 20.0 28859.0 19.0 4.0 null 3.0 2012-03-09T00:00:00.000+0000 371.0 1668486 140570353 E03 14019363 Medical Incident 02/26/2014 02/26/2014 2014-02-26T19:23:49.000+0000 2014-02-26T19:26:40.000+0000 2014-02-26T19:26:51.000+0000 2014-02-26T19:28:37.000+0000 2014-02-26T19:34:11.000+0000 null null Patient Declined Transport 2014-02-26T19:54:26.000+0000 1500 Block of SUTTER ST SF 94109.0 B04 38 3324.0 2 2 2 true Non Life-threatening 1 ENGINE 1.0 4 5 Pacific Heights (37.787081788666, -122.425867820141) 140570353-E03 11.0 13.0 6.0 30.0 28858.0 27.0 9.0 null null 2014-02-26T00:00:00.000+0000 622.0 1668487 140570353 84 14019363 Medical Incident 02/26/2014 02/26/2014 2014-02-26T19:23:49.000+0000 2014-02-26T19:26:40.000+0000 2014-02-26T19:26:51.000+0000 2014-02-26T19:27:07.000+0000 null null null Patient Declined Transport null 1500 Block of SUTTER ST SF 94109.0 B04 38 3324.0 2 2 2 true Non Life-threatening 1 MEDIC 2.0 4 5 Pacific Heights (37.787081788666, -122.425867820141) 140570353-84 11.0 13.0 6.0 30.0 28858.0 27.0 9.0 null null 2014-02-26T00:00:00.000+0000 null 1584518 141691481 T09 14058327 Structure Fire 06/18/2014 06/18/2014 2014-06-18T12:23:45.000+0000 2014-06-18T12:24:32.000+0000 2014-06-18T12:24:39.000+0000 2014-06-18T12:26:11.000+0000 null null null Fire 2014-06-18T12:29:40.000+0000 2800 Block of HARRISON ST San Francisco 94110.0 B06 07 5533.0 3 3 3 false Alarm 1 TRUCK 7.0 6 9 Mission (37.751835432252

<b> Finding: </b> After examining some CALL NUMBERS that have more than 1 row in the data set, it is clear that this is NOT a data quality issue. Instead, some emergency calls might request more than one UNIT to help. In another word, this data set contains 50536 unit dispatches for 49948 calls

<b> 2.3 There are several columns related to neighorbood, let us take a look first </b>

In [0]:
%sql
select distinct `Neighborhooods - Analysis Boundaries`, `Neighborhoods - Analysis Boundaries`, Neighborhoods
from sparkView
order by `Neighborhooods - Analysis Boundaries`

Neighborhooods - Analysis Boundaries Neighborhoods - Analysis Boundaries Neighborhoods Bayview Hunters Point 1.0 1.0 Bernal Heights 2.0 2.0 Castro/Upper Market 5.0 3.0 Chinatown 6.0 4.0 Excelsior 7.0 5.0 Financial District/South Beach 8.0 6.0 Glen Park 10.0 7.0 Golden Gate Park 12.0 8.0 Haight Ashbury 3.0 9.0 Hayes Valley 9.0 10.0 Inner Richmond 11.0 11.0 Inner Sunset 14.0 12.0 Japantown 15.0 13.0 Lakeshore 16.0 14.0 Lincoln Park 17.0 15.0 Lone Mountain/USF 18.0 16.0 Marina 13.0 17.0 McLaren Park 19.0 18.0 Mission 20.0 19.0 Mission Bay 4.0 20.0 Nob Hill 21.0 21.0 Noe Valley 22.0 22.0 None null null North Beach 23.0 23.0 Oceanview/Merced/Ingleside 24.0 24.0 Outer Mission 28.0 25.0 Outer Richmond 29.0 26.0 Pacific Heights 30.0 27.0 Pacific Heights 30.0 17.0 Portola 25.0 28.0 Potrero Hill 26.0 29.0 Potrero Hill 20.0 19.0 Presidio 27.0 30.0 Presidio Heights 31.0 31.0 Russian Hill 32.0 32.0 Seacliff 33.0 33.0 South of Market 34.0 34.0 Sunset/Parkside 35.0 35.0 Sunset/Parkside 35.0 8.0 Tenderloin 36.0 36.0 Treasure Island 37.0 37.0 Twin Peaks 38.0 38.0 Visitacion Valley 40.0 39.0 West of Twin Peaks 41.0 40.0 Western Addition 39.0 41.0

In [0]:
%sql
select distinct `Neighborhooods - Analysis Boundaries`
from sparkView
order by `Neighborhooods - Analysis Boundaries`

Neighborhooods - Analysis Boundaries Bayview Hunters Point Bernal Heights Castro/Upper Market Chinatown Excelsior Financial District/South Beach Glen Park Golden Gate Park Haight Ashbury Hayes Valley Inner Richmond Inner Sunset Japantown Lakeshore Lincoln Park Lone Mountain/USF Marina McLaren Park Mission Mission Bay Nob Hill Noe Valley None North Beach Oceanview/Merced/Ingleside Outer Mission Outer Richmond Pacific Heights Portola Potrero Hill Presidio Presidio Heights Russian Hill Seacliff South of Market Sunset/Parkside Tenderloin Treasure Island Twin Peaks Visitacion Valley West of Twin Peaks Western Addition

<p> Note that The San Francisco Planning Department officially identifies 36 neighborhoods buut this data set contains 41 neighborhoods with distinct names.</p>
https://data.sfgov.org/Geographic-Locations-and-Boundaries/Planning-Neighborhood-Groups-Map/iacs-ws63

for the sake of time, I will skip the investigation why we have more than 36 neighborhoods

# Question 1: What are the date ranges for this dataset?
Answer: 2000-04-13 to 2019-08-31（Please find SQL query below）

In [0]:
%sql
select count(*) from sparkView where Call_Date_Dt is null;

count(1) 0

In [0]:
%sql
select min(Call_Date_Dt), max(Call_Date_Dt) from sparkView;

min(Call_Date_Dt) max(Call_Date_Dt) 2000-04-13T00:00:00.000+0000 2019-08-31T00:00:00.000+0000

# Question 2: What are the neighborhoods the fire department serves?
Answer: 41 neighborhoods are served by the fire departments.

Note that there is a NONE value in the related column which is not a valid neighborhood name

In [0]:
%sql
select distinct `Neighborhooods - Analysis Boundaries`
  from sparkView
  where `Neighborhooods - Analysis Boundaries` != 'None'
  order by `Neighborhooods - Analysis Boundaries`

Neighborhooods - Analysis Boundaries Bayview Hunters Point Bernal Heights Castro/Upper Market Chinatown Excelsior Financial District/South Beach Glen Park Golden Gate Park Haight Ashbury Hayes Valley Inner Richmond Inner Sunset Japantown Lakeshore Lincoln Park Lone Mountain/USF Marina McLaren Park Mission Mission Bay Nob Hill Noe Valley North Beach Oceanview/Merced/Ingleside Outer Mission Outer Richmond Pacific Heights Portola Potrero Hill Presidio Presidio Heights Russian Hill Seacliff South of Market Sunset/Parkside Tenderloin Treasure Island Twin Peaks Visitacion Valley West of Twin Peaks Western Addition

# Question 3: What’s the average delay (defined as the difference between a call’s receive time and the on scene time) for all calls in 2019? What’s the 99%-tile delay?
Answer: average delay: 586 seconds,  99%-tile delay 2294 seconds (Please find the SQL query below)

Please note that multiple units might be dispatched for the same call and each unit have its own on scene time, the statistics are computed based on <b> all the distinct dispatched units instead of all the calls </b>.

In [0]:
%sql
select count(*) from sparkView
  where year(Call_Date_Dt) == 2019

count(1) 2085

In [0]:
%sql
select count(*) from sparkView
  where year(Call_Date_Dt) == 2019 and isnull(`On Scene DtTm`)

count(1) 369

In [0]:
%sql
select count(*) from sparkView
  where year(Call_Date_Dt) == 2019 and isnull(`Received DtTm`)

count(1) 0

In [0]:
%sql
select min(Delay), percentile(Delay, 0.01), mean(Delay), percentile(Delay, 0.99), max(Delay), count(*)
  from sparkView 
  where year(Call_Date_Dt) == 2019 and !isnull(`On Scene DtTm`)

min(Delay) percentile(Delay, CAST(0.01 AS DOUBLE), 1) mean(Delay) percentile(Delay, CAST(0.99 AS DOUBLE), 1) max(Delay) count(1) 0.0 0.0 586.0285547785547 2294.2999999999984 6135.0 1716

Note that there are some calls that has ZERO delay which could be a data quality issue? For the sake of time, I won't further investigate into it

In [0]:
%sql
select *
  from sparkView 
  where year(Call_Date_Dt) == 2019 and !isnull(`On Scene DtTm`) and delay == 0

col1 Call Number Unit ID Incident Number Call Type Call Date Watch Date Received DtTm Entry DtTm Dispatch DtTm Response DtTm On Scene DtTm Transport DtTm Hospital DtTm Call Final Disposition Available DtTm Address City Zipcode of Incident Battalion Station Area Box Original Priority Priority Final Priority ALS Unit Call Type Group Number of Alarms Unit Type Unit sequence in call dispatch Fire Prevention District Supervisor District Neighborhooods - Analysis Boundaries Location RowID Supervisor Districts Fire Prevention Districts Current Police Districts Neighborhoods - Analysis Boundaries Zip Codes Neighborhoods Police Districts Civic Center Harm Reduction Project Boundary HSOC Zones Call_Date_Dt Delay 178161 190370121 76 19015356 Medical Incident 02/06/2019 02/05/2019 2019-02-06T01:22:18.000+0000 2019-02-06T01:22:18.000+0000 2019-02-06T01:22:18.000+0000 2019-02-06T01:22:18.000+0000 2019-02-06T01:22:18.000+0000 null null Patient Declined Transport 2019-02-06T01:55:57.000+0000 MISSION ST/17TH ST San Francisco 94110.0 B02 07 5236.0 2 2 2 true Non Life-threatening 1 MEDIC 1.0 2 9 Mission (37.763429274013, -122.419512736792) 190370121-76 7.0 8.0 7.0 20.0 28859.0 19.0 4.0 null 3.0 2019-02-06T00:00:00.000+0000 0.0 127698 190950485 AM102 19040357 Medical Incident 04/05/2019 04/04/2019 2019-04-05T05:47:19.000+0000 2019-04-05T05:47:19.000+0000 2019-04-05T05:47:19.000+0000 2019-04-05T05:47:19.000+0000 2019-04-05T05:47:19.000+0000 2019-04-05T05:52:26.000+0000 2019-04-05T06:08:35.000+0000 Code 2 Transport 2019-04-05T06:57:55.000+0000 25TH ST/GUERRERO ST San Francisco 94110.0 B06 11 5535.0 2 2 2 false Non Life-threatening 1 PRIVATE 1.0 6 8 Mission (37.75037224151, -122.422725612019) 190950485-AM102 5.0 2.0 7.0 20.0 28859.0 19.0 4.0 null null 2019-04-05T00:00:00.000+0000 0.0 104212 191224245 KM07 19052073 Medical Incident 05/02/2019 05/02/2019 2019-05-02T23:16:33.000+0000 2019-05-02T23:16:33.000+0000 2019-05-02T23:16:33.000+0000 2019-05-02T23:16:33.000+0000 2019-05-02T23:16:33.000+0000 2019-05-02T23:26:07.000+0000 2019-05-02T23:48:41.000+0000 Code 2 Transport 2019-05-03T00:12:00.000+0000 POLK ST/EDDY ST San Francisco 94109.0 B02 03 3114.0 2 2 2 false Non Life-threatening 1 PRIVATE 1.0 2 6 Tenderloin (37.78310139927685, -122.4191817050345) 191224245-KM07 9.0 7.0 6.0 36.0 28852.0 36.0 9.0 1.0 null 2019-05-02T00:00:00.000+0000 0.0 143794 190753642 E06 19032234 Medical Incident 03/16/2019 03/16/2019 2019-03-16T20:23:11.000+0000 2019-03-16T20:23:11.000+0000 2019-03-16T20:23:11.000+0000 2019-03-16T20:23:11.000+0000 2019-03-16T20:23:11.000+0000 null null SFPD 2019-03-16T20:33:34.000+0000 16TH ST/MARKET ST San Francisco 94114.0 B05 06 5252.0 3 3 3 true Potentially Life-Threatening 1 ENGINE 1.0 5 8 Castro/Upper Market (37.764195714116, -122.433074471116) 190753642-E06 5.0 15.0 7.0 5.0 28862.0 3.0 4.0 null 5.0 2019-03-16T00:00:00.000+0000 0.0 143805 190753568 E44 19032226 Medical Incident 03/16/2019 03/16/2019 2019-03-16T20:04:05.000+0000 2019-03-16T20:04:05.000+0000 2019-03-16T20:04:05.000+0000 2019-03-16T20:04:05.000+0000 2019-03-16T20:04:05.000+0000 null null Code 2 Transport 2019-03-16T20:14:11.000+0000 SAN BRUNO AV/PAUL AV San Francisco 94134.0 B10 44 6326.0 3 3 3 true Potentially Life-Threatening 1 ENGINE 1.0 10 9 Portola (37.723919996468, -122.402060968175) 190753568-E44 7.0 10.0 3.0 25.0 309.0 28.0 3.0 null null 2019-03-16T00:00:00.000+0000 0.0 134483 190864254 AP 19036938 Other 03/27/2019 03/27/2019 2019-03-27T23:23:17.000+0000 2019-03-27T23:23:17.000+0000 2019-03-27T23:23:17.000+0000 2019-03-27T23:23:17.000+0000 2019-03-27T23:23:17.000+0000 null null Fire 2019-03-27T23:23:17.000+0000 CALL BOX: SF INTERNATIONAL AIRPORT null null B09 44 6913.0 3 3 3 false Alarm 1 AIRPORT 1.0 None None None (37.616882323925, -122.384094238098) 190864254-AP null null null null 60.0 null null null null 2019-03-27T00:00:00.000+0000 0.0 158682 190590701 86 19024929 Medical Incident 02/28/2019 02/28/2019 2019-02-28T08:34:22.000+0000 2019-02-28T08:34:22.000+0000 201

# Question 4: Are the number of calls trending up, or trending down, on a year-over-year basis?
Answer: Overall speaking, number of calls are trending up, except a sudden increase in 2009 and then decrease in 2010. (Pleaes find the corresponding visualization below)

Note that this does not taking year 2010 and 2019 into account because we don't have data for all the months for those two years.

In [0]:
%sql
create table table_distinct_calls as
  select distinct `Call Number` as Call_Number, Call_Date_Dt, year(Call_Date_Dt) as year
  from sparkView

num_affected_rows num_inserted_rows

In [0]:
%sql
select year, count(*) from table_distinct_calls
  group by year
  order by year

year count(1) 2000 1578 2001 2189 2002 2236 2003 2352 2004 2359 2005 2259 2006 2360 2007 2312 2008 2593 2009 2385 2010 2506 2011 2620 2012 2668 2013 2693 2014 2858 2015 2886 2016 2960 2017 3009 2018 3060 2019 2065

< < < THE END > > >